feature/experimental/gribi/gribi.go

```diff
@@ -0,0 +1,209 @@
+package gribi
+
+import (
+	"context"
+	"reflect"
+	"sync"
+	"testing"
+	"time"
+	"unsafe"
+
+	grbp "github.com/openconfig/gribi/v1/proto/service"
+	"github.com/openconfig/gribigo/chk"
+	"github.com/openconfig/gribigo/constants"
+	"github.com/openconfig/gribigo/fluent"
+	"github.com/openconfig/ondatra"
+	"google.golang.org/grpc"
+)
+
+const (
+	timeout   = time.Minute
+	maxUint64 = 1<<64 - 1
+)
+
+// GRIBI provides access to GRIBI APIs of the DUT.
+type GRIBIHandler struct {
+	dut            *ondatra.DUTDevice
+	gribiC         grbp.GRIBIClient
+	fluentC        *fluent.GRIBIClient
+	fibAck         bool
+	persistence    bool
+	redundancyMode fluent.RedundancyMode
+	lastElectionID *grbp.Uint128
+}
+
+var mu sync.Mutex
+
+func NewGRIBIFluent(t testing.TB, dut *ondatra.DUTDevice, persistance, fibAck bool) *GRIBIHandler {
+	fluentClient := fluent.NewClient()
+	gribiC := dut.RawAPIs().GRIBI().New(t)
+	fluentClient.Connection().WithStub(gribiC)
+	g := &GRIBIHandler{dut: dut,
+		gribiC:         gribiC,
+		fluentC:        fluentClient,
+		fibAck:         fibAck,
+		persistence:    persistance,
+		redundancyMode: fluent.ElectedPrimaryClient, // all of test cases use ElectedPrimaryClient
+		lastElectionID: &grbp.Uint128{Low: 1,
+			High: 0,
+		},
+	}
+	if persistance {
+		fluentClient.Connection().WithInitialElectionID(g.lastElectionID.Low, g.lastElectionID.High).WithRedundancyMode(fluent.ElectedPrimaryClient).WithPersistence()
+	} else {
+		fluentClient.Connection().WithInitialElectionID(g.lastElectionID.Low, g.lastElectionID.High).WithRedundancyMode(fluent.ElectedPrimaryClient)
+	}
+	if fibAck {
+		fluentClient.Connection().WithFIBACK()
+	}
+	fluentClient.Start(context.Background(), t)
+	fluentClient.StartSending(context.Background(), t)
+	err := g.awaitTimeout(context.Background(), t, timeout)
+	if err != nil {
+		t.Fatalf("can not establish the session: %v", err)
+	}
+	return g
+}
+
+func (g *GRIBIHandler) Close(t testing.TB) {
+	t.Helper()
+	t.Logf("closing GRIBI connection for dut: %s", g.dut.Name())
+	g.fluentC.Stop(t)
+	v := reflect.Indirect(reflect.ValueOf(g.gribiC)).FieldByName("cc")
+	clientConn, ok := (reflect.NewAt(v.Type(), unsafe.Pointer(v.UnsafeAddr())).Elem().Interface()).(*grpc.ClientConn)
```

Comment:
I see what the problem is. The Ondatra API only gives us the client stub, and it is not designed to close the underlying connection if it's new. For now, let's use the Default() stub, but consider filing an issue with Ondatra to rectify the RawAPIs design.

Accessing unexported fields using reflection is problematic.

Reply:
@liulk  using  Default() is not going to help here since Ondatra does not provide close for default gribi client as well.
I have discussed this with @greg-dennis (here)[https://github.com/openconfig/ondatra/pull/3] and he suggested using reflection. Anyway, If you agree we have this code here until we finalize the solution for this.

Reply:
Yeah, I suggested getting the connection field by reflection because the need to close the connection seemed like a rather rare requirement, making it hard to justify the additional weight and complexity on the binding that would be needed to support it. We also have no resources to design the changes to accommodate the use case at this time, and so far this is the one and only request for that feature. I don't love it, but I also don't see the generated proto code changing in a way to break it any time soon.

Two suggestions:
(1) I suspect "reflect.Indirect(reflect.ValueOf(g.gribiC))" would be better as "reflect.ValueOf(g.gribiC).Elem()"
(2) Does every test really need to close the connection? I thought it was a fairly narrow test case that actually needed the connection closed. Might be better to confine this to just that test case rather than making it part of a common library. The above-mentioned gribigo/fluent library has found no need to ever close the connection.

Reply:
You can definitely reuse the gRIBI client stub over many fluent clients. The abstraction that defines a gRIBI session (election ID, persistence) is contained entirely within the Modify streaming RPC.

Hence, `RawAPIs().GRIBI().Default()` is the better way. There is no need to close this default transport since it's just in the background and indefinitely reusable.

Reply:
Thanks for the clarification!  I changed the code as you suggested. There is no need to use reflection.